src/common/Formatter.cc

```diff
@@ -293,7 +291,9 @@ void JSONFormatter::dump_int(const char *name, int64_t s)
 
 void JSONFormatter::dump_float(const char *name, double d)
 {
-  add_value(name, d);
+  char foo[30];
+  snprintf(foo, sizeof(foo), "%lf", d);
```

Comment:
i think the reason we have `0.8100000000000001` when printing `810 / 1000.0` is just that IEEE754 is not able represent all decimal numbers with a fraction part. the default precision of `printf()` is 6. that's why we have `0.810000` if we print the number using `printf()` without specifying the number of digits. if we instead use
```c++
char foo[30];
double f = 810 / 1000.0;
snprintf(foo, sizeof(foo), "%.*f", std::numeric_limits<decltype(f)>::max_digits10, f);
cout << foo << endl;
```
we will have exactly the same output as
```c++
cout.precision(std::numeric_limits<double>::max_digits10);
cout << f << endl;
```

but i understand your concern -- a nice looking decimal number is printed as if it has had very small fraction part. but the reality is that it does =(
the default precision of 6 just hides the ugly fact.

Reply:
I'm not really sure what the issue is with high precision numbers in JSON output which is not meant to be human consumable. @dzafman is this really a problem?

Reply:
@batrick @tchaikov I'm keeping everything in microseconds which could also be considered milliseconds in a fix point representation with scale 1000 (3 digits).  So I'm thinking of replacing the N / 1000.0 with something that creates a string from an integer and the decimal scale as a number of digits.  There's a little leading 0 stuff to handle too.  

1234 -> 1.234
234 -> 0.234
2 -> 0.002

To me this is human consumable from the command line

ceph daemon mgr.x dump_osd_network

This corresponds to new ceph health output which is using default float stream precision:

# ceph health
OSD_SLOW_PING_TIME_FRONT Long heartbeat ping times on front interface seen, longest is 1.372 msec
    Slow heartbeat ping on front interface from osd.0 to osd.2 1.372 msec

# ceph daemon mgr.x dump_osd_network
...
        {
            "last update": "Wed Aug  7 19:51:30 2019",
            "stale": false,
            "from osd": 0,
            "to osd": 2,
            "interface": "front",
            "average": {
                "1min": 1.242,
                "5min": 1.3720000000000001,
                "15min": 1.345
            },
 ...

Reply:
@batrick @tchaikov I'm going to avoid using dump_float(), so I don't care if we close this pull request.  I think we should lower the precision by 1 (or 2) from the maximum so that rounding hides the issue with representation of decimal numbers.

Reply:
Ceph utils should give `plain` format by default. JSON output should be printed only if requested explicitly. So, it's better to change default format for this command.